src/combinatorics/simple_graph/coloring.lean

```diff
@@ -0,0 +1,403 @@
+    { rintro v w hvw,
+      simp only [complete_graph_eq_top, top_adj, subtype.mk_eq_mk, ne.def],
+      exact C.valid hvw, } }
+end
+
+lemma colorable_set_nonempty_of_colorable {n : ℕ} (hc : G.colorable n) :
+  {n : ℕ | G.colorable n}.nonempty :=
+⟨n, hc⟩
+
+lemma chromatic_number_bdd_below : bdd_below {n : ℕ | G.colorable n} :=
+⟨0, λ _ _, zero_le _⟩
+
+lemma chromatic_number_le [fintype α] (C : G.coloring α) :
+  G.chromatic_number ≤ fintype.card α :=
+cInf_le chromatic_number_bdd_below C.to_colorable
+
+lemma colorable_chromatic_number {m : ℕ} (hc : G.colorable m) :
+  G.colorable G.chromatic_number :=
+begin
+  dsimp only [chromatic_number],
+  rw nat.Inf_def,
+  apply nat.find_spec,
+  exact colorable_set_nonempty_of_colorable hc,
+end
+
+lemma colorable_chromatic_number_of_fintype (G : simple_graph V) [fintype V] :
+  G.colorable G.chromatic_number :=
+colorable_chromatic_number G.colorable_of_fintype
+
+lemma chromatic_number_le_one_of_subsingleton (G : simple_graph V) [subsingleton V] :
+  G.chromatic_number ≤ 1 :=
+begin
+  rw chromatic_number,
+  apply cInf_le chromatic_number_bdd_below,
+  fsplit,
+  refine coloring.mk (λ _, 0) _,
+  intros v w,
+  rw subsingleton.elim v w,
+  simp,
+end
+
+lemma chromatic_number_eq_zero_of_isempty (G : simple_graph V) [is_empty V] :
+  G.chromatic_number = 0 :=
+begin
+  rw ←nonpos_iff_eq_zero,
+  apply cInf_le chromatic_number_bdd_below,
+  apply colorable_of_is_empty,
+end
+
+lemma is_empty_of_chromatic_number_eq_zero (G : simple_graph V) [fintype V]
+  (h : G.chromatic_number = 0) : is_empty V :=
+begin
+  have h' := G.colorable_chromatic_number_of_fintype,
+  rw h at h',
+  exact G.is_empty_of_colorable_zero h',
+end
+
+lemma zero_lt_chromatic_number [nonempty V] {n : ℕ} (hc : G.colorable n) :
+  0 < G.chromatic_number :=
+begin
+  apply le_cInf (colorable_set_nonempty_of_colorable hc),
+  intros m hm,
+  by_contra h',
+  simp only [not_le, nat.lt_one_iff] at h',
+  subst h',
+  obtain ⟨i, hi⟩ := hm.some (classical.arbitrary V),
+  exact nat.not_lt_zero _ hi,
+end
+
+lemma colorable_lower_bound {G' : simple_graph V} (h : G ≤ G') (n : ℕ) (hc : G'.colorable n) :
+  G.colorable n :=
+⟨hc.some.comp (hom.map_spanning_subgraphs h)⟩
+
+lemma chromatic_number_le_of_forall_imp {G' : simple_graph V}
+  {m : ℕ} (hc : G'.colorable m)
+  (h : ∀ n, G'.colorable n → G.colorable n) :
+  G.chromatic_number ≤ G'.chromatic_number :=
+begin
+  apply cInf_le chromatic_number_bdd_below,
+  apply h,
+  apply colorable_chromatic_number hc,
+end
+
+lemma chromatic_number_lower_bound (G' : simple_graph V)
+  {m : ℕ} (hc : G'.colorable m) (h : G ≤ G') :
+  G.chromatic_number ≤ G'.chromatic_number :=
+begin
+  apply chromatic_number_le_of_forall_imp hc,
+  exact colorable_lower_bound h,
+end
+
+lemma chromatic_number_le_n_of_colorable {n : ℕ} (hc : G.colorable n) :
+  G.chromatic_number ≤ n :=
+begin
+  rw chromatic_number,
+  apply cInf_le chromatic_number_bdd_below,
+  fsplit,
+  exact classical.choice hc,
+end
```

Comment:
Could you move this lemma as far up as it can go? It's an important one, and I'm surprised to find it so far down!